InnerEye/ML/lightning_models.py

```diff
@@ -442,22 +481,33 @@ def training_or_validation_step(self,
         logits = self.model(*model_inputs_and_labels.model_inputs)
         loss = self.loss_fn(logits, labels)
         self.write_loss(is_training, loss)
-        metrics = create_metrics_dict_for_scalar_models(is_classification_model=self.is_classification_model,
-                                                        sequence_target_positions=self.sequence_target_positions)
-        compute_scalar_metrics(metrics,
-                               subject_ids=model_inputs_and_labels.subject_ids,
-                               model_output=self.logits_to_posterior(logits),
-                               labels=labels,
-                               loss_type=self.loss_type)
-        for hue, name, value in metrics.enumerate_single_values():
-            hue_suffix = "" if hue == MetricsDict.DEFAULT_HUE_KEY else "/" + hue
-            self.log_on_epoch(name=name + hue_suffix, value=value, is_training=is_training)
+        self.compute_and_log_metrics(logits, labels, is_training)
         self.log_on_epoch(name=MetricType.SUBJECT_COUNT,
                           value=len(model_inputs_and_labels.subject_ids),
                           is_training=is_training,
                           reduce_fx=sum)
+        self.log_time_per_batch(is_training)
         return loss
 
+    def compute_and_log_metrics(self, logits: torch.Tensor, targets: torch.Tensor, is_training: bool):
+        metrics = self.train_metrics_dict if is_training else self.val_metrics_dict
+        for i, (hue, metric_list) in enumerate(metrics.items()):
+            hue_suffix = "" if hue == MetricsDict.DEFAULT_HUE_KEY else f"/{hue}"
+            # mask the model outputs and labels if required
+            masked_model_outputs_and_labels = get_masked_model_outputs_and_labels(
+                logits[:, i, ...], targets[:, i, ...])
+            # compute metrics on valid masked tensors only
+            if masked_model_outputs_and_labels is not None:
+                _model_output, _labels = \
+                    masked_model_outputs_and_labels.model_outputs.data, \
+                    masked_model_outputs_and_labels.labels.data
+                # Convert labels to the same datatype as the model outputs, necessary when running with AMP
+                _labels = _labels.to(dtype=_model_output.dtype)
```

Comment:
do we still need that? We've gotten rid of a lot of the explicit conversion code, and left that all to PL.

Reply:
Good question...

Reply:
Nope